Advanced IADS (Skynet)

RndName opened the pull request:
The current skynet implementation was completly rewritten and factored out to a class which handles the iads_network within the conflict theater. So units will not be handled according to theire name (e.g. RED|SAM...) anymore. They will be included into the mission via the dcsliberation lua table which the skynet script can iterate through.

This will implement #324

(reopened from #1994)

### The IADS Network
The The IADS Network class will handle connections between all elements of the IADS Network:
- Sam, SamAsEWR
- EWR (including Naval)
- PointDefence of Sams (e.g. SA-15 / Tor for high value sites like S-300 or similar). PointDefense units can be defined inside a template. As example i added them to S-300 and Patriot sites already.
- **new** Connection Nodes
- **new** Power Sources
- **new** Command Centers

All of the IADS Buildings make use of the layout system of course. So they can also have different variations.

Documentation already available over here: https://github.com/dcs-liberation/dcs_liberation/wiki/Custom-Campaigns#iads---integrated-air-defence-system

This new system also allows scenery map objects to be part of the IADS System using TriggerZones. To let skynet use these objects a dummy unit will be placed on top which allows skynet to work. Important: It has to be ensured by campaign designers that these objects can be destroyed within the game. I tried VOR / Tacans as comm tower but they can not be destroyed within the game.. Also real power sources on the map can be used.

### Basic and Advanced Mode
This new system can be used in two modes:
- **Basic**: General usage like in the current system without power sources, connection nodes and command centers. All IADS elements can always communicate with each other / participate to the IADS without the need to link them.
- **Advanced**: Support for Connection Nodes, Power Sources and command centers. All Elements require a working power source and a connection with other elements to participate to the IADS Network. If they loose the connection they work autonomous. If they loose power they go dark. Command Centers are used as the central nervous system. If it gets destroyed the whole system works autonomous.

### Implementation inside the campaign
The advanced system can be implemented by campaign designers with two approaches:
1. campaign designers can define the iads completly in the campaign yaml ("By Config")
2. campaign designers can just place the ground units without a definition in the campaign yaml. Then it will be automatically calculated ("By Range")

Example for a config:

```yaml
version: "10.1" # Campaign Version 10.1 allows the advanced iads network config
advanced_iads: true # Campaign has connection_nodes / power_sources / command_centers
iads_config:
  - Aircraft Carrier # A naval group participating to the IADS network without any connections
  - Naval Group # Naval Group
  - Golan North-100: # Group name of the SAM, EWR or Command Center used in the campaign.miz
    - IADS_CN2 # All connected nodes (Comms, Power, PointDefense)
    - IADS_PS4 # Exact name of the connected group used in the campaign.miz
    - IADS_PS3
  - Golan North-11:
    - IADS_CN2
    - IADS_PS3
  - Golan North-139:
    - IADS_PS2
    - IADS_CN6
  ...
```

This information will then be parsed by a special crafted IADS Network Class which generates all the connections and will later export them in a LUA Table which can be parsed by the skynet script.

Example of how the advanced system will look using the map:
![grafik](https://user-images.githubusercontent.com/6678803/152533258-b51a0ce9-5958-4dd2-a60d-cdaa3467fc97.png)

[Documentation](https://github.com/RndName/dcs_liberation/blob/advanced-skynet/doc/iads/advanced_iads.md)

### Open Tasks
- [x] Update to the lates changes
- [x] Rework the map implementation to fit the recent changes
- [x] Finish the map (correct color and so on)
- [x] Check and Correct basic iads mode (pd should not be added for example)
- [x] Add Ships & Carriers as EWRs
- [ ] Hide IADS Layer on map when no advanced_iads is active (maybe just hide with css?)
- [ ] Complete the on tgo update routine for the map layer (it gets updated but the lines need to be redrawn)
- [x] Finish Documentation (https://github.com/dcs-liberation/dcs_liberation/wiki/Custom-Campaigns#iads---integrated-air-defence-system)
- [x] Fix AI issues with SEAD mission when using skynet #2141 
- [x] Update Skynet to V3.0.0 (important: this is a develop version - not oficially released yet)
- [x] Testing.
- [x] Skynet property overrides for specific settings (also describe in the wiki)

RndName committed:
Refactor luagenerator

- cleaned up the generation
- created special class to handle the serialization
- improved string escaping: Replace OS Path separator with normal slash and allow the usage of a single quote in unit names by changing the delimiter to double quote instead (1797)
- adjusted unit_name generation to prevent scripting errors with unescaped characters

RndName committed:
Implement advanced skynet functions

- factor out own class for the iadsnetwork within the conflicttheater
- This class will handle all Skynet related things - no specific group_name handling necessary in future
- make iadsbuilding own TGO class because SAM & EWRs are Vehicle Groups. IADS Elements dont have any groups attached.
- added command center, connection node and power source as Ground objects which can be added by the campaign designer
- adjust lua generator to support new iads units
- parse the campaign yaml to get the iads network information
- use the range as fallback if no yaml information was found
- complete rewrite of the skynet lua script
- allow destruction of iads network to be persistent over all rounds
- modified the presetlocation handling: the wrapper PresetLocation for PointWithHeading now stores the original name from the campaign miz to have the ability to process campaign yaml configurations based on the ground unit
- Implementation of the UI representation for the IADS Network
- Give user the option to enable or disable advanced iads
- Extended the layout system: Implement Sub task handling to support PD

RndName committed:
Update Golan Heights to support advanced skynet

- Updated to 10.1 (headings + advanced iads)
- Demonstrates how the advanced system will work

RndName committed:
Bump campaign version to 10.2

RndName committed:
Fix incorrect faction loading

RndName committed:
Update Mist and Skynet scripts (V3.0.0)

- adopt skynet config to v3

kevotosmart commented:
Quick question in regards to this. 
> * **new** Connection Nodes


How would those be defined in a campaign ?

RndName commented:
> How would those be defined in a campaign ?

@kevotosmart you can find the current documentation over here: https://github.com/dcs-liberation/dcs_liberation/wiki/Custom-Campaigns#iads---integrated-air-defence-system
They will be defined as any other ground object (By placing one of the placeholder units in the campaign.miz or by using map objects), For comms it would be the `Comms tower M` unit

RndName committed:
Allow skynet properties override and improve dead unit handling

Now the user can override special skynet properties from the unit.yaml of the main Radar of the sam site. Which unit needs these overrides can be looked up over here: https://github.com/walder/Skynet-IADS/blob/develop/skynet-iads-source/skynet-iads-supported-types.lua

- fixed wrong exclusion of groups with the first unit dead but still able to participate
- added some exception handling in the lua script for groups which skynet is not able to control. preventing scripting errors
